### ISSUE TITLE
Multiple SQL correlations requests launch simultaneously

### ISSUE THREAD

**Issue by ater49** [MISP/MISP#4801]
As told to @iglocska during FIRST, we have an issue during attributes re-correlation work.

### Work environment

| Questions                 | Answers
|---------------------------|--------------------
| Type of issue             | Bug
| OS version (server)       | Ubuntu 16.04.6 LTS
| PHP version               | PHP 7.0.33-0ubuntu0.16.04.5
| MISP version / git hash   | 2.4.109
|RAM| 2 Go
|CPU|4 core

### Expected behavior
MISP wait previous correlation request end to start a new one. 

### Actual behavior
Multiple SQL correlations requests are simultaneously launching for the same event and new other correlation request are launching without waiting other correlations request end.

### Steps to reproduce the behavior
Set MISP.completely_disable_correlation to False


### Logs, screenshots, configuration dump, ...

|  11 | misp | localhost | misp | Query   |   74 | Sending data | SELECT `Correlation`.`event_id` FROM `misp`.`correlations` AS `Correlation`   WHERE `Correlation`.`1_event_id` = (29405)  GROUP BY `Correlation`.`event_id`  ORDER BY `Correlation`.`event_id` DESC                                                                                                                                                                                                                                                                                                                                                                                                                                                         |    0.000 |
|  69 | misp | localhost | misp | Query   |   74 | Sending data | SELECT `Correlation`.`event_id` FROM `misp`.`correlations` AS `Correlation`   WHERE `Correlation`.`1_event_id` = (29405)  GROUP BY `Correlation`.`event_id`  ORDER BY `Correlation`.`event_id` DESC                                                                                                                                                                                                                                                                                                                                                                                                                                                         |    0.000 |
| 112 | misp | localhost | misp | Query   |   62 | Sending data | SELECT `Correlation`.`event_id` FROM `misp`.`correlations` AS `Correlation`   WHERE `Correlation`.`1_event_id` = (29405)  GROUP BY `Correlation`.`event_id`  ORDER BY `Correlation`.`event_id` DESC                                                                                                                                                                                                                                                                                                                                                                                                                                                         |    0.000 |
| 155 | misp | localhost | misp | Query   |   62 | Sending data | SELECT `Correlation`.`event_id` FROM `misp`.`correlations` AS `Correlation`   WHERE `Correlation`.`1_event_id` = (29405)  GROUP BY `Correlation`.`event_id`  ORDER BY `Correlation`.`event_id` DESC                                                                                                                                                                                                                                                                                                                                                                                                                                                         |    0.000 |
| 195 | misp | localhost | misp | Query   |   74 | Sending data | SELECT `Correlation`.`event_id` FROM `misp`.`correlations` AS `Correlation`   WHERE `Correlation`.`1_event_id` = (29405)  GROUP BY `Correlation`.`event_id`  ORDER BY `Correlation`.`event_id` DESC                                                                                                                                                                                                                                                                                                                                                                                                                                                         |    0.000 |
| 234 | misp | localhost | misp | Query   |   74 | Sending data | SELECT `Correlation`.`event_id` FROM `misp`.`correlations` AS `Correlation`   WHERE `Correlation`.`1_event_id` = (29405)  GROUP BY `Correlation`.`event_id`  ORDER BY `Correlation`.`event_id` DESC                                                                                                                                                                                                                                                                                                                                                                                                                                                         |    0.000 |
| 269 | misp | localhost | misp | Query   |   62 | Sending data | SELECT `Correlation`.`event_id` FROM `misp`.`correlations` AS `Correlation`   WHERE `Correlation`.`1_event_id` = (29405)  GROUP BY `Correlation`.`event_id`  ORDER BY `Correlation`.`event_id` DESC                                                                                                                                                                                                                                                                                                                                                                                                                                                         |    0.000 |
| 308 | misp | localhost | misp | Query   |   74 | Sending data | SELECT `Correlation`.`event_id` FROM `misp`.`correlations` AS `Correlation`   WHERE `Correlation`.`1_event_id` = (29405)  GROUP BY `Correlation`.`event_id`  ORDER BY `Correlation`.`event_id` DESC                                                                                                                                                                                                                                                                                                                                                                                                                                                         |    0.000 |

**Comment by lu-chi** [comment]
Hello there. Was that already fixed? I have similar issue. Is there a workaround or a way to fix that - at least temporarily to allow MISP to perform on accepted level?

**Comment by mokaddem** [comment]
I see that there is an option in the MISP config `MISP.deadlock_avoidance` that is used for the correlations but I'm not entirely sure if it could fix the issue described here.
Maybe @iglocska could shed light on this option?

**Comment by lu-chi** [comment]
Thanks for info @mokaddem. Would be nice to have a solution for that as this seriously impacts the instances I work with. Correlations queries needs to be killed to allow the user to log in and then to work inside MISP (website). If not - every single action is blocked by that what is causing a lot of issues. @iglocska would that be possible to described in details all steps to solve the case? (obviously if there is a solution and case is a known one).
Thanks in advance!

**Comment by iglocska** [comment]
I have not seen this reproduced just yet on our end, could you describe the scenario that causes this? Are you hammering your API with a lot of edit requests? If so, can you share the scripts that you are using? (Full event edits continuously could be refactored into per attribute queries for example).

@mokaddem the setting you mentioned will simply convert some bulk queries into atomic ones, turning a few long queries into many short ones. This should help with deadlock avoidance but will generally increase query times altogether.

**Comment by lu-chi** [comment]
@iglocska thanks for reply. What I mean by the correlation is it simply blocks the users while login and working with website frontend. The only solution I was able to provide was to find a MariaDB process and simply kill that one (the processes constantly commin up are correlation table queries). Once killed it allows the user to login or to work within MISP. However these processes are spawning up again once killed so the procedure needs to be repated a few times to stop it for a longer period of time. I am not sure why the correlation query is blocking the user and entire MISP for the time it is executed but definitely something is wrong over there. I would hihly appreciate help in that.
